stdlib/LinearAlgebra/src/abstractq.jl

```diff
@@ -0,0 +1,628 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+abstract type AbstractQ{T} end
+
+struct AdjointQ{T,S<:AbstractQ{T}} <: AbstractQ{T}
+    Q::S
+end
+
+parent(adjQ::AdjointQ) = adjQ.Q
+eltype(::Type{<:AbstractQ{T}}) where {T} = T
+ndims(::AbstractQ) = 2
+
+# inversion/adjoint/transpose
+inv(Q::AbstractQ) = Q'
+adjoint(Q::AbstractQ) = AdjointQ(Q)
+transpose(Q::AbstractQ{<:Real}) = AdjointQ(Q)
+transpose(Q::AbstractQ) = error("transpose not implemented for $(typeof(Q)). Consider using adjoint instead of transpose.")
+adjoint(adjQ::AdjointQ) = adjQ.Q
+
+# promotion with AbstractMatrix, at least for equal eltypes
+promote_rule(::Type{<:AbstractMatrix{T}}, ::Type{<:AbstractQ{T}}) where {T} =
+    (@inline; Union{AbstractMatrix{T},AbstractQ{T}})
+
+# conversion
+AbstractQ{S}(Q::AbstractQ{S}) where {S} = Q
+# the following eltype promotion needs to be defined for each subtype
+# convert(::Type{AbstractQ{T}}, Q::QType) where {T} = QType{T}(Q)
+convert(::Type{AbstractQ{T}}, Q::AbstractQ{T}) where {T} = Q
```

Comment:
Hm, that's annoying. It seems that this "fallback" doesn't kick in for specific subtypes of `AbstractQ`?